src/FakeItEasy.Analyzer/UnusedReturnValueAnalyzer.cs

```diff
@@ -0,0 +1,87 @@
+namespace FakeItEasy.Analyzer
+{
+    using System;
+    using System.Collections.Immutable;
+    using System.Linq;
+    using Microsoft.CodeAnalysis;
+    using Microsoft.CodeAnalysis.CSharp;
+    using Microsoft.CodeAnalysis.CSharp.Syntax;
+    using Microsoft.CodeAnalysis.Diagnostics;
+    using static DiagnosticDefinitions;
+
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class UnusedReturnValueAnalyzer : DiagnosticAnalyzer
+    {
+        private static readonly ImmutableDictionary<string, DiagnosticDescriptor> Diagnostics =
+            GetDiagnosticsMap(
+                nameof(UnusedCallSpecification));
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = Diagnostics.Values.ToImmutableArray();
+
+        public override void Initialize(AnalysisContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            context.RegisterSyntaxNodeAction(AnalyzeCall, SyntaxKind.InvocationExpression);
+        }
+
+        private static void AnalyzeCall(SyntaxNodeAnalysisContext context)
+        {
+            var call = context.Node as InvocationExpressionSyntax;
+            if (call == null)
+            {
+                return;
+            }
+
+            var methodSymbol = GetCalledMethodSymbol(call, context);
+            if (methodSymbol == null)
+            {
+                return;
+            }
+
+            var diagnosticName = GetDiagnosticName(methodSymbol);
+            if (diagnosticName == null)
+            {
+                return;
+            }
+
+            var descriptor = Diagnostics.GetValueOrDefault(diagnosticName);
+            if (descriptor == null)
+            {
+                return;
+            }
+
+            if (call.Parent is StatementSyntax)
+            {
+                var firstArg = call.ArgumentList.Arguments.FirstOrDefault();
+                var lambda = firstArg?.Expression as LambdaExpressionSyntax;
+                var callSpec = lambda?.Body?.ToString()
+                               ?? firstArg?.ToString()
+                               ?? "(unknown)";
+                var diagnostic = Diagnostic.Create(descriptor, call.GetLocation(), callSpec);
+                context.ReportDiagnostic(diagnostic);
+            }

```

Comment:
I just realized that this part of the code isn't really generic; it's specific to `A.CallTo`, since it tries to extract the call specification. I could probably refactor further to have some diagnostic-specific code.